### PR TITLE
MAINT: array types: rename `as_xparray` to `_asarray`

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
Towards gh-18866, https://github.com/scipy/scipy/issues/18866#issuecomment-1634998051

#### What does this implement/fix?
- Rename `as_xparray` to `_asarray`. This reflects its temporary nature, as things we would like to see standardised into `xp.asarray` (or otherwise become available in the standard).
- Add some `array_namespace` calls for consistency
- Add some `xp=xp` arguments to reduce redundant calls to `array_namespace`
- Update the docs accordingly
- Document the `copy` helper

#### Additional information
I also removed some misleading lines in the docs, see gh-19799. This can be replaced with something more robust at a later point (but no need for now, in most cases `array_namespace` is all you need).

### CODE REVIEW

File: scipy/cluster/hierarchy.py [hunk 1]
```diff
@@ -2330,7 +2330,8 @@ def num_obs_linkage(Z):
     12
 
     """
-    Z = as_xparray(Z, order='c')
+    xp = array_namespace(Z)
+    Z = _asarray(Z, order='c', xp=xp)
```
Comment (lucascolley) [inline scipy/cluster/hierarchy.py:2334]: The only benefit of making this a one-liner is LoC - the downsides are that it is less explicit, and the LoC will return if we are able to use `xp.asarray` anyway. Furthermore, if the function was changed to later use `xp`, this change would have to be made (else have duplicate `array_namespace` calls).

**Comment by rgommers** [comment]
Thanks Lucas!

> Rename `as_xparray` to `_asarray`. This reflects its temporary nature, as things we would like to see standardised into `xp.asarray` (or otherwise become available in the standard).

I'm fine with the `_asarray` name (it's a bit of a bike shed to me, the current name works too), however it seems very unlikely that `order` and `check_finite` keywords will be standardized. It's also not necessary I think. Does this change your mind on anything in this PR?

**Comment by rgommers** [comment]
That sounds about right. Okay, I'll see about merging this tomorrow if there are no other comments.

**Review by rgommers — APPROVED** [review]
Okay, in it goes!

**Pull request merged**